File: scripts/solana/test_phased.mjs
#!/usr/bin/env node
/**
 * End-to-end test for UltraHonk verification using the SDK
 * 
 * This is the single test entrypoint - it dogfoods our own SDK.
 */
import { Connection, Keypair, PublicKey } from '@solana/web3.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Import our SDK
import { SolanaNoirVerifier, PROOF_SIZE, VK_SIZE } from '../../sdk/dist/index.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RPC_URL = process.env.RPC_URL || 'http://127.0.0.1:8899';
const PROGRAM_ID = new PublicKey(process.env.PROGRAM_ID || '7sfMWfVs6P1ACjouyvRwWHjiAj6AsFkYARP2v9RBSSoe');
const CIRCUIT_NAME = process.env.CIRCUIT || 'simple_square';
const DEBUG = process.env.DEBUG === '1';

// Circuit paths
const proofPath = path.join(__dirname, `../../test-circuits/${CIRCUIT_NAME}/target/keccak/proof`);
const piPath = path.join(__dirname, `../../test-circuits/${CIRCUIT_NAME}/target/keccak/public_inputs`);
const vkPath = path.join(__dirname, `../../test-circuits/${CIRCUIT_NAME}/target/keccak/vk`);

async function main() {
    console.log('\n╔══════════════════════════════════════════════════════════════╗');
    console.log('║         UltraHonk Verification Test (using SDK)              ║');
    console.log('╚══════════════════════════════════════════════════════════════╝\n');
    
    console.log(`Circuit: ${CIRCUIT_NAME}`);
    console.log(`Program: ${PROGRAM_ID.toBase58()}`);
    console.log(`RPC: ${RPC_URL}\n`);
    
    // Load test data
    if (!fs.existsSync(proofPath)) {
        console.error(`❌ Proof not found: ${proofPath}`);
        console.error(`   Run: cd test-circuits/${CIRCUIT_NAME} && ./build.sh`);
        process.exit(1);
    }
    
    const proof = fs.readFileSync(proofPath);
    const publicInputsRaw = fs.readFileSync(piPath);
    const vk = fs.readFileSync(vkPath);
    
    // Convert raw PI bytes to array of 32-byte buffers
    const numPi = publicInputsRaw.length / 32;
    const publicInputs = [];
    for (let i = 0; i < numPi; i++) {
        publicInputs.push(publicInputsRaw.slice(i * 32, (i + 1) * 32));
    }
    
    console.log(`Proof: ${proof.length} bytes (expected: ${PROOF_SIZE})`);
    console.log(`VK: ${vk.length} bytes (expected: ${VK_SIZE})`);
    console.log(`Public inputs: ${numPi} × 32 = ${publicInputsRaw.length} bytes\n`);
    
    // Setup
    const connection = new Connection(RPC_URL, 'confirmed');
    const payer = Keypair.generate();
    
    // Fund payer
    console.log('Funding payer account...');
    const airdropSig = await connection.requestAirdrop(payer.publicKey, 10_000_000_000);
    await connection.confirmTransaction(airdropSig);
    console.log('  ✅ Funded\n');
    
    // Create SDK client
    const verifier = new SolanaNoirVerifier(connection, { programId: PROGRAM_ID });
    
    // =========================================================================
    // STEP 1: Upload VK (one-time per circuit)
    // =========================================================================
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║           CIRCUIT DEPLOYMENT (one-time per circuit)           ║');
    console.log('╚══════════════════════════════════════════════════════════════╝');
    
    const vkStart = Date.now();
    const vkResult = await verifier.uploadVK(payer, vk);
    const vkTime = Date.now() - vkStart;
    
    console.log(`  VK Account: ${vkResult.vkAccount.toBase58()}`);
    console.log(`  Chunks: ${vkResult.numChunks}`);
    console.log(`  Time: ${(vkTime / 1000).toFixed(2)}s`);
    console.log('  ✅ VK uploaded\n');
    
    // =========================================================================
    // STEP 2: Verify proof
    // =========================================================================
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║              PROOF VERIFICATION (per proof)                   ║');
    console.log('╚══════════════════════════════════════════════════════════════╝');
    
    const verifyStart = Date.now();
    let lastPhase = '';
    
    const result = await verifier.verify(payer, proof, publicInputs, vkResult.vkAccount, {
        onProgress: (phase, current, total) => {
            if (phase !== lastPhase) {
                if (lastPhase) console.log('  ✅ Done');
                console.log(`  ${phase}...`);
                lastPhase = phase;
            }
        }
    });
    if (lastPhase) console.log('  ✅ Done');
    
    const verifyTime = Date.now() - verifyStart;
    
    console.log(`\n  Verified: ${result.verified ? '✅ YES' : '❌ NO'}`);
    console.log(`  Total CUs: ${result.totalCUs.toLocaleString()}`);
    console.log(`  Transactions: ${result.numTransactions} (${result.numSteps} sequential steps)`);
    console.log(`  Time: ${(verifyTime / 1000).toFixed(2)}s\n`);
    
    if (!result.verified) {
        console.error('❌ Verification failed!');
        process.exit(1);
    }
    
    // =========================================================================
    // STEP 3: Create receipt (for integrators)
    // =========================================================================
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║         RECEIPT CREATION (for integrators)                    ║');
    console.log('╚══════════════════════════════════════════════════════════════╝');
    
    const [receiptPda] = verifier.deriveReceiptPda(vkResult.vkAccount, publicInputs);
    console.log(`  Receipt PDA: ${receiptPda.toBase58()}`);
    
    await verifier.createReceipt(
        payer,
        result.stateAccount,
        result.proofAccount,
        vkResult.vkAccount,
        publicInputs
    );
    console.log('  ✅ Receipt created\n');
    
    // =========================================================================
    // STEP 4: Validate receipt (simulating integrator lookup)
    // =========================================================================
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║         RECEIPT VALIDATION (integrator check)                 ║');
    console.log('╚══════════════════════════════════════════════════════════════╝');
    
    const receipt = await verifier.getReceipt(vkResult.vkAccount, publicInputs);
    
    if (receipt) {
        console.log(`  ✅ Receipt found!`);
        console.log(`     Verified at slot: ${receipt.verifiedSlot}`);
        console.log(`     Verified at: ${new Date(Number(receipt.verifiedTimestamp) * 1000).toISOString()}`);
    } else {
        console.log('  ❌ Receipt not found');
        process.exit(1);
    }
    
    // =========================================================================
    // SUMMARY
    // =========================================================================
    console.log('\n╔══════════════════════════════════════════════════════════════╗');
    console.log('║                         SUMMARY                               ║');
    console.log('╚══════════════════════════════════════════════════════════════╝');
    
    console.log(`\n  Circuit: ${CIRCUIT_NAME}`);
    console.log(`  VK deployment: ${(vkTime / 1000).toFixed(2)}s (one-time)`);
    console.log(`  Proof verification: ${(verifyTime / 1000).toFixed(2)}s`);
    console.log(`  Total CUs: ${result.totalCUs.toLocaleString()}`);
    console.log(`  Transactions: ${result.numTransactions} (${result.numSteps} sequential steps)`);
    console.log(`\n  VK Account: ${vkResult.vkAccount.toBase58()}`);
    console.log(`  Receipt: ${receiptPda.toBase58()}`);
    
    console.log('\n  🎉 All tests passed!\n');
}

main().catch(err => {
    console.error('\n❌ Test failed:', err.message);
    if (DEBUG) console.error(err.stack);
    process.exit(1);
});

File: crates/rust-sdk/examples/test_phased.rs
//! End-to-end test for UltraHonk verification using the Rust SDK
//!
//! This mirrors the JavaScript test_phased.mjs - it's the main test entrypoint
//! for the Rust SDK.
//!
//! Usage:
//!   cargo run --example test_phased                    # Test simple_square (default)
//!   CIRCUIT=merkle_membership cargo run --example test_phased  # Test specific circuit
//!   VERBOSE=1 cargo run --example test_phased          # Show detailed output
//!
//! Environment variables:
//!   RPC_URL     - RPC endpoint (default: http://127.0.0.1:8899)
//!   PROGRAM_ID  - Verifier program ID (default: uses surfnet deployed program)
//!   CIRCUIT     - Circuit to test (default: simple_square)
//!   VERBOSE     - Show detailed output (default: 0)

use solana_client::rpc_client::RpcClient;
use solana_commitment_config::CommitmentConfig;
use solana_noir_verifier_sdk::{
    ReceiptInfo, SolanaNoirVerifier, VerificationResult, VerifierConfig, VerifyOptions,
    VkUploadResult, PROOF_SIZE, VK_SIZE,
};
use solana_sdk::{
    native_token::LAMPORTS_PER_SOL, pubkey::Pubkey, signature::Keypair, signer::Signer,
};
use std::{
    env, fs,
    path::PathBuf,
    str::FromStr,
    sync::Arc,
    time::{Duration, Instant},
};

fn main() {
    env_logger::init();

    // Configuration from environment
    let rpc_url = env::var("RPC_URL").unwrap_or_else(|_| "http://127.0.0.1:8899".to_string());
    let program_id_str = env::var("PROGRAM_ID")
        .unwrap_or_else(|_| "7sfMWfVs6P1ACjouyvRwWHjiAj6AsFkYARP2v9RBSSoe".to_string());
    let circuit_name = env::var("CIRCUIT").unwrap_or_else(|_| "simple_square".to_string());
    let _verbose = env::var("VERBOSE").unwrap_or_else(|_| "0".to_string()) == "1";

    let program_id = Pubkey::from_str(&program_id_str).expect("Invalid PROGRAM_ID");

    println!("\n╔══════════════════════════════════════════════════════════════╗");
    println!("║       UltraHonk Verification Test (Rust SDK)                 ║");
    println!("╚══════════════════════════════════════════════════════════════╝\n");

    println!("Circuit: {}", circuit_name);
    println!("Program: {}", program_id);
    println!("RPC: {}", rpc_url);

    // Load test data
    let circuit_paths = get_circuit_paths(&circuit_name);
    let proof = fs::read(&circuit_paths.proof).unwrap_or_else(|_| {
        panic!(
            "Proof not found: {:?}\n   Run: cd test-circuits/{} && ./build.sh",
            circuit_paths.proof, circuit_name
        )
    });
    let public_inputs =
        fs::read(&circuit_paths.public_inputs).expect("Failed to read public inputs");
    let vk = fs::read(&circuit_paths.vk).expect("Failed to read VK");

    println!("\nProof: {} bytes (expected: {})", proof.len(), PROOF_SIZE);
    println!("VK: {} bytes (expected: {})", vk.len(), VK_SIZE);
    println!(
        "Public inputs: {} × 32 = {} bytes\n",
        public_inputs.len() / 32,
        public_inputs.len()
    );

    // Setup client
    let client = Arc::new(RpcClient::new_with_commitment(
        rpc_url.clone(),
        CommitmentConfig::confirmed(),
    ));

    // Generate and fund payer
    let payer = Keypair::new();
    println!("Funding payer account...");

    // Request airdrop
    let airdrop_sig = client
        .request_airdrop(&payer.pubkey(), 10 * LAMPORTS_PER_SOL)
        .expect("Airdrop failed");

    // Wait for confirmation
    for _ in 0..30 {
        std::thread::sleep(Duration::from_millis(500));
        if let Ok(Some(result)) = client.get_signature_status(&airdrop_sig) {
            if result.is_ok() {
                break;
            }
        }
    }
    println!("  ✅ Funded\n");

    // Create verifier
    let verifier = SolanaNoirVerifier::new(client.clone(), VerifierConfig::new(program_id));

    // VK Upload
    println!("╔══════════════════════════════════════════════════════════════╗");
    println!("║           CIRCUIT DEPLOYMENT (one-time per circuit)          ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    let vk_start = Instant::now();
    let vk_result: VkUploadResult = verifier.upload_vk(&payer, &vk).expect("VK upload failed");
    let vk_time = vk_start.elapsed();

    println!("  VK Account: {}", vk_result.vk_account);
    println!("  Chunks: {}", vk_result.num_chunks);
    println!("  Time: {:.2}s", vk_time.as_secs_f64());
    println!("  ✅ VK uploaded\n");

    // Verification
    println!("╔══════════════════════════════════════════════════════════════╗");
    println!("║              PROOF VERIFICATION (per proof)                  ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    let verify_start = Instant::now();
    let result: VerificationResult = verifier
        .verify(
            &payer,
            &proof,
            &public_inputs,
            &vk_result.vk_account,
            Some(VerifyOptions::new().without_auto_close()), // Don't auto-close, we'll do it manually
        )
        .expect("Verification failed");
    let verify_time = verify_start.elapsed();

    println!(
        "\n  Verified: {}",
        if result.verified { "✅ YES" } else { "❌ NO" }
    );
    println!("  Total CUs: {}", format_number(result.total_cus));
    println!(
        "  Transactions: {} ({} sequential steps)",
        result.num_transactions, result.num_steps
    );
    println!("  Time: {:.2}s", verify_time.as_secs_f64());

    if !result.verified {
        eprintln!("\n❌ Verification failed!");
        std::process::exit(1);
    }

    // Receipt
    println!("\n╔══════════════════════════════════════════════════════════════╗");
    println!("║         RECEIPT CREATION (for integrators)                   ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    let (receipt_pda, _) = verifier.derive_receipt_pda(&vk_result.vk_account, &public_inputs);
    println!("  Receipt PDA: {}", receipt_pda);

    verifier
        .create_receipt(
            &payer,
            &result.state_account,
            &result.proof_account,
            &vk_result.vk_account,
            &public_inputs,
        )
        .expect("Receipt creation failed");
    println!("  ✅ Receipt created\n");

    // Validate receipt
    println!("╔══════════════════════════════════════════════════════════════╗");
    println!("║         RECEIPT VALIDATION (integrator check)                ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    let receipt: Option<ReceiptInfo> = verifier
        .get_receipt(&vk_result.vk_account, &public_inputs)
        .expect("Failed to get receipt");

    if let Some(info) = receipt {
        println!("  ✅ Receipt found!");
        println!("     Verified at slot: {}", info.verified_slot);
        println!(
            "     Verified at: {}",
            format_timestamp(info.verified_timestamp)
        );
    } else {
        eprintln!("  ❌ Receipt not found");
        std::process::exit(1);
    }

    // Close accounts
    println!("\n╔══════════════════════════════════════════════════════════════╗");
    println!("║         ACCOUNT CLEANUP (recover rent)                       ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    let (recovered, _) = verifier
        .close_accounts(&payer, &result.state_account, &result.proof_account)
        .expect("Failed to close accounts");
    let sol_recovered = recovered as f64 / LAMPORTS_PER_SOL as f64;
    println!("  ✅ Accounts closed");
    println!(
        "     Recovered: {:.4} SOL ({} lamports)",
        sol_recovered, recovered
    );

    // Summary
    println!("\n╔══════════════════════════════════════════════════════════════╗");
    println!("║                         SUMMARY                              ║");
    println!("╚══════════════════════════════════════════════════════════════╝");

    println!("\n  Circuit: {}", circuit_name);
    println!("  VK deployment: {:.2}s (one-time)", vk_time.as_secs_f64());
    println!("  Proof verification: {:.2}s", verify_time.as_secs_f64());
    println!("  Total CUs: {}", format_number(result.total_cus));
    println!(
        "  Transactions: {} ({} sequential steps)",
        result.num_transactions, result.num_steps
    );
    println!("  Rent recovered: {:.4} SOL", sol_recovered);
    println!("\n  VK Account: {}", vk_result.vk_account);
    println!("  Receipt: {}", receipt_pda);

    println!("\n  🎉 All tests passed!\n");
}

struct CircuitPaths {
    proof: PathBuf,
    public_inputs: PathBuf,
    vk: PathBuf,
}

fn get_circuit_paths(circuit_name: &str) -> CircuitPaths {
    // Find workspace root
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let workspace_root = PathBuf::from(manifest_dir)
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .to_path_buf();

    let circuit_dir = workspace_root
        .join("test-circuits")
        .join(circuit_name)
        .join("target")
        .join("keccak");

    CircuitPaths {
        proof: circuit_dir.join("proof"),
        public_inputs: circuit_dir.join("public_inputs"),
        vk: circuit_dir.join("vk"),
    }
}

fn format_number(n: u64) -> String {
    let s = n.to_string();
    let mut result = String::new();
    for (i, c) in s.chars().rev().enumerate() {
        if i > 0 && i % 3 == 0 {
            result.insert(0, ',');
        }
        result.insert(0, c);
    }
    result
}

fn format_timestamp(ts: i64) -> String {
    use std::time::{SystemTime, UNIX_EPOCH};

    if ts <= 0 {
        return "N/A".to_string();
    }

    let dt = UNIX_EPOCH + Duration::from_secs(ts as u64);
    match dt.duration_since(UNIX_EPOCH) {
        Ok(_) => {
            // Simple ISO-ish format
            let now = SystemTime::now();
            match now.duration_since(UNIX_EPOCH) {
                Ok(now_dur) => {
                    let diff = now_dur.as_secs().saturating_sub(ts as u64);
                    if diff < 60 {
                        format!("{} seconds ago", diff)
                    } else if diff < 3600 {
                        format!("{} minutes ago", diff / 60)
                    } else {
                        format!("Unix timestamp: {}", ts)
                    }
                }
                Err(_) => format!("Unix timestamp: {}", ts),
            }
        }
        Err(_) => format!("Unix timestamp: {}", ts),
    }
}
